### ISSUE TITLE
Feature Request: Default pod affinity 

### ISSUE THREAD

**Issue by jhamman** [dask/dask-kubernetes#146]
Dask-kubernetes is currently configured to include two default pod/taint tolerations (see #109, #110, #145). This feature allows dask-worker pods to be the sole inhabitants of a node/node-pool, provided the cluster is configured with the corresponding taints. Another scheduling pattern would be to attract dask-worker pods to a specific set of nodes. While this has been done successfully with nodeSelectors in the past, it is not a particularly user-friendly option.

I'm wondering if we could come up with a universal dask-worker label and then have dask-kubernetes set a `preferredDuringSchedulingIgnoredDuringExecution` affinity on each pod. Has anyone tried this? Are there specific downsides to doing this? My understanding is that the nodeSelector approach will be eventually deprecated in favor of affinity rules.

I think something like this would work (haven't tested):

```yaml
spec:
  affinity:
    nodeAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
      - weight: 100
        preference:
          matchExpressions:
          - key: k8s.dask.org/node-purpose
            operator: In
            values:
            - worker
``` 

The main benefits here are:

1. This doesn't require users/admins to set anything, existing applications without labels would continue to function as normal
2. It would allow admins to direct traffic to specific nodes without asking the user to configure their pod template

Thoughts from @yuvipanda, @jacobtomlinson, @betatim?

**Comment by jhamman** [comment]
For reference, the KubeSpawer includes the following affinity rule for single user notebook pods:
```yaml
  affinity:
    nodeAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
      - preference:
          matchExpressions:
          - key: hub.jupyter.org/node-purpose
            operator: In
            values:
            - user
        weight: 100
```

**Comment by consideRatio** [comment]
If I understand you correctly @jhamman, you are proposing to add a default preferred nodeAffinity to the worker pods created by dask-kubernetes. This default affinity would relate to a specific dask label. I like that, and see no practical downside to doing so.

**Comment by consideRatio** [comment]
### About the Z2JH default nodeAffinity
KubeSpawner get's configured by the Z2JH Helm chart / jupyterhub_config.py to have that affinity preference by default, but it can be configured to be required or to have no such affinity at all.

See: http://z2jh.jupyter.org/en/latest/reference.html#scheduling-userpods-nodeaffinity
And: http://z2jh.jupyter.org/en/latest/reference.html#scheduling-corepods-nodeaffinity

---

### About my use of the Dask helm chart (but not dask-kubernetes)
I'm using the Dask helm chart to get a dask scheduler and dask workers, but not dask-kubernetes. I have configured my Dask scheduler to live on a "core node pool" and the workers on a preemptible node pool.

The `nodeAffinity` can configure _preferredDuringScheduling..._ as well as _requiredDuringScheduling..._.

With the requiredDuringScheduling... you can accomplish the same and more than you can with nodeSelector, but nodeSelector remains and is quick and easy to use. I'm also thinking it could be deprecated in the future as it can be replaced fully already.

```yaml
# just for reference, some yaml representing parts of my my helm values provided to
# the Dask helm chart. It that creates a scheduler and workers pods for me.
# I have also made it use the same image as my z2jh user environment
scheduler:
  nodeSelector:
    hub.jupyter.org/node-purpose: core

worker:
  tolerations:
    - key: "dedicated"
      operator: "Equal"
      value: "dask-workers"
      effect: "NoSchedule"
  nodeSelector:
    node-purpose: dask-workers
```

**Comment by jhamman** [comment]
@consideRatio - 

> If I understand you correctly @jhamman 

Yes, you've got it. 

> With the requiredDuringScheduling... you can accomplish the same and more than you can with nodeSelector, but nodeSelector remains and is quick and easy to use. I'm also thinking it could be deprecated in the future as it can be replaced fully already.

My understanding is that nodeSelector is likely to go away in favor of nodeAffinity. Is this what you meant?

**Comment by jhamman** [comment]
I've opened https://github.com/dask/dask-kubernetes/pull/147 to help this conversation.

**Comment by consideRatio** [comment]
That is what I meant!

---

I noted [in the docs](http://kubernetes.dask.org/en/latest/#deployment-details), that it was stated:

> If you have nodes with the corresponding taints, then the worker pods will schedule to those nodes (and no other pods will be able to schedule to those nodes).

This may be what happens practically a lot as there will be an empty node that only these can schedule on, they would certainly be able to schedule elsewhere and would prefer to schedule on a less busy node if there was such node available. Unless... They had a nodeAffinity of the "preferredDuring..." kind, then they would opt for the tainted node they tolerade unless it was full but would not trigger a scale up if they still could schedule elsewhere. But, if they have a "requiredDuring..." kind of nodeAffinity for the tainted node that they tolerate, then they would also trigger a node scale up if they could not schedule there.

**Comment by jhamman** [comment]
Okay, I see where you are headed. I think we should start with a default prefer affinity. Then follow up with making the prefer vs require option configurable at the dask kubernetes config level.

**Comment by jhamman** [comment]
closed by #147
